### PR TITLE
chore: check for non-labeled issues

### LINKED ISSUE #16273: Run a GitHub Action weekly to check for issues that are missing labels
## Intro

I wrote some GitHub issue search queries that help us check if any issues are wrongly labeled.

Maybe we can automate this task, by writing a GitHub Action script.

## Requirements

1. Action runs weekly
1. Check if any queries have a match
    - If a query matches (so an issue is missing labels), throw error
    - If no query matches, exit cleanly

### Queries

```markdown
https://github.com/renovatebot/renovate/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+-label%3Astatus%3Arequirements+-label%3Astatus%3Aready+-label%3Astatus%3Ain-progress+-label%3Astatus%3Ablocked+-label%3Astatus%3Awaiting-on-response+

https://github.com/renovatebot/renovate/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+-label%3Atype%3Abug+label%3Astatus%3Aready+-label%3Atype%3Afeature+-label%3Atype%3Adocs+-label%3Atype%3Arefactor+

https://github.com/renovatebot/renovate/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+-label%3Apriority-1-critical+-label%3Apriority-2-high+-label%3Apriority-3-medium+-label%3Apriority-4-low++-label%3Apriority-5-triage
```

## Todo:

- [ ] Put "type of issue" query in action
- [ ] Put "priority query" in action
- [ ] Run action weekly
- [ ] Make action error if any query matches
- [ ] Else action exits cleanly
- [ ] If action works properly, update [`docs/development/issue-labeling`](https://github.com/renovatebot/renovate/blob/main/docs/development/issue-labeling.md) to mention the GitHub action check run

### PR DESCRIPTION
**Pull request by ph1l1pF**
## Changes

This PR will create a new workflow that once in a week will check all open issues and search if any of them are not labeled correctly. If all of them are labeled as expected, it'll cleanly complete the job, otherwise, a new issue reporting the unlabeled issues will be created.

I have tested the action on a test repository: https://github.com/ph1l1pF/renovate-test/issues

## Context

Closes #16273 

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [x] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — COMMENTED** [review]
Can we determine what the results would be for this repo prior to merging? e.g. via a console output or running the same logic manually locally?

File: .github/workflows/check-unlabeled-issues.yml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+name: 'Check unlabeled issues'
```
Comment (HonkingGoose) [inline .github/workflows/check-unlabeled-issues.yml:1]:
```suggestion
name: 'Find issues with missing labels'
```

This explains better what the Action will do, and will search for.

Also, rename the `.yml` file so it matches the new name.

File: .github/workflows/check-unlabeled-issues.yml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  check-unlabeled-issues:
+    runs-on: ubuntu-latest
+
+    env:
+      GH_TOKEN: ${{ github.token }}
+
+    steps:
+      - uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
+
+      - name: Check for unlabeled issues
```
Comment (HonkingGoose) [inline .github/workflows/check-unlabeled-issues.yml:21]:
```suggestion
      - name: Search for issues with missing labels
```

File: docs/development/issue-labeling.md [hunk 3]
```diff
@@ -221,3 +221,11 @@ Add a label `auto:retry-latest` to any Discussion where the user should retry th
...
 
 Apply the `self-hosted` label when an issue is applicable only to users who self-administer their own bot.
+
+## Automated label report
+
+We have a GitHub Action that checks if we're using the correct labels on our issues.
+
+The action runs once a week and checks for issues without the proper labels.
+
+If the action finds any issues missing labels, it creates a new issue with the missing labels and details the issues found so any collaborator can label them accordingly.
```
Comment (HonkingGoose) [inline docs/development/issue-labeling.md:231]:
```suggestion
## Automated check for Issues with missing labels

We have a GitHub Action (`find-issues-with-missing labels.yml`) to find issues on our repository that are missing labels.
Any Issues with missing labels will be put in a list in a new "error" Issue.

The Action runs each week.

### Apply the correct labels manually

The Action will not fix any badly labeled issues.
This means that you, or we, must apply the correct labels to any affected Issue.
```

File: tools/check-unlabeled-issues.sh [hunk 4]
```diff
@@ -0,0 +1,78 @@
+#!/bin/bash
```
Comment (HonkingGoose) [inline tools/check-unlabeled-issues.sh:1]: I will review the script a bit later.

**Review by HonkingGoose — COMMENTED** [review]
First round of review.

I'll review the Bash script soon. 😉

File: tools/check-unlabeled-issues.sh [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  # Create a new label "Label check action"
+  gh label create "$LABEL_CHECK_ACTION" --description "This label is used for the issue that lists invalid issues." --repo $REPO || { echo "Failed to create label"; exit 1; }
+fi
+
+LABEL_CHECK_ISSUE_EXISTS=$(gh search issues --label "$LABEL_CHECK_ACTION" --repo $REPO --json number) || { echo "Failed to fetch existing label check issue"; exit 1; }
+ISSUE_NUMBER=$(echo "$LABEL_CHECK_ISSUE_EXISTS" | jq -r '.[].number')
+
+if [ -z "$ISSUE_NUMBER" ]; then
+
+  # Create a new issue with the list of issues
```
Comment (HonkingGoose) [inline tools/check-unlabeled-issues.sh:61]:
```suggestion
  # Create a new issue (with the list of issues in it).
```

File: tools/check-unlabeled-issues.sh [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+  gh issue create --repo $REPO --label "$LABEL_CHECK_ACTION" --title "$ISSUE_TITLE" --body "$(echo -e "$ISSUE_BODY")" || { echo "Failed to create issue"; exit 1; }
+else
+  # Update the existing issue with the list of issues
+  gh issue edit "$ISSUE_NUMBER" --repo $REPO --title "$ISSUE_TITLE" --body "$(echo -e "$ISSUE_BODY")" || { echo "Failed to update issue"; exit 1; }
+
+  # Reopen the issue
+  gh issue reopen "$ISSUE_NUMBER" --repo $REPO || { echo "Failed to reopen issue"; exit 1; }
+fi
+
+# Provide an output in the action itself
```
Comment (HonkingGoose) [inline tools/check-unlabeled-issues.sh:71]:
```suggestion
# Show the list of "issues with missing labels" in the logs.
```

Please check if this is correct! 😉

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
Second round of reviews, this time for the _comments_ in the script.

After the suggestions and changes have been made, we need to check the technical bits of the Bash script itself.

File: tools/check-unlabeled-issues.sh (hunk 6 above)
Comment (ph1l1pF) [reply tools/check-unlabeled-issues.sh:71]: correct :)

File: tools/check-unlabeled-issues.sh [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+  gh label create "$LABEL_CHECK_ACTION" --description "This label is used for the issue that lists invalid issues." --repo $REPO || { echo "Failed to create label"; exit 1; }
+fi
+
+LABEL_CHECK_ISSUE_EXISTS=$(gh search issues --label "$LABEL_CHECK_ACTION" --repo $REPO --json number) || { echo "Failed to fetch existing label check issue"; exit 1; }
+ISSUE_NUMBER=$(echo "$LABEL_CHECK_ISSUE_EXISTS" | jq -r '.[].number')
+
+if [ -z "$ISSUE_NUMBER" ]; then
+
+  # Create a new issue with the list of issues
+  gh issue create --repo $REPO --label "$LABEL_CHECK_ACTION" --title "$ISSUE_TITLE" --body "$(echo -e "$ISSUE_BODY")" || { echo "Failed to create issue"; exit 1; }
```
[the remark this answers is not in the document]
Comment (ph1l1pF) [reply tools/check-unlabeled-issues.sh:62]: add point to sentence

**Comment by ph1l1pF** [comment]
> ve been made, we need to check the technical bits of the Bash script itself.

@HonkingGoose comment resolved

**Review by rarkins — COMMENTED** [review]
Could you explain what the `Label check action` label is used for? And in particular: is this solution impossible without it?

**Comment by rarkins** [comment]
> Could you explain what the `Label check action` label is used for? And in particular: is this solution impossible without it?

Could the issue be identified instead using author + issue title?

**Comment by ph1l1pF** [comment]
> > Could you explain what the `Label check action` label is used for? And in particular: is this solution impossible without it?
> 
> Could the issue be identified instead using author + issue title?

@rarkins The label is just a way to identify the issue that lists the other issues. Title + author (bot) ist also possible. This would have the benefit that no label has to be created. I can try that.

**Comment by ph1l1pF** [comment]
> > > Could you explain what the `Label check action` label is used for? And in particular: is this solution impossible without it?
> > 
> > 
> > Could the issue be identified instead using author + issue title?
> 
> @rarkins The label is just a way to identify the issue that lists the other issues. Title + author (bot) ist also possible. This would have the benefit that no label has to be created. I can try that.

@rarkins I have done it.

**Review by rarkins — DISMISSED** [review]

**Comment by rarkins** [comment]
# Label check action
## Found 12 issues missing `type:` labels:
- #33087
- #32830
- #32083
- #28925
- #27543
- #27349
- #27295
- #27034
- #27032
- #26571
- #23009
- #5607
## Found 3 issues missing `priority:` labels:
- #33087
- #31504
- #29420

File: tools/find-issues-with-missing-labels.sh [hunk 8]
```diff
@@ -39,21 +39,11 @@ for FILTER in "$TYPE_LABELS_FILTER" "$PRIORITY_LABELS_FILTER"; do
 done
 
 if [ "$HAS_ISSUES_MISSING_LABELS" = false ]; then
-  echo "Found no issues with missing labels. Exiting the action."
+  echo "Found no issues with missing labels. Exiting the action with success."
```
Comment (HonkingGoose) [inline tools/find-issues-with-missing-labels.sh:42]:
```suggestion
  echo "All checked issues have labels. Exiting the action."
```

File: tools/find-issues-with-missing-labels.sh [hunk 9]
```diff
@@ -72,7 +62,7 @@ fi
 echo -e "$ISSUE_BODY"
 
 # Log a message and "fail" the Action if there are issues with missing labels
-echo "Found issues missing labels. Please check the issue(s) above."
+echo "Found issues missing labels. Please check the issue(s) above. Exiting the action with failure."
```
Comment (HonkingGoose) [inline tools/find-issues-with-missing-labels.sh:65]:
```suggestion
echo "Found issues without labels. Please check the issue(s) listed above. Exiting the action."
```

File: docs/development/issue-labeling.md [hunk 10]
```diff
@@ -221,3 +221,15 @@ Add a label `auto:retry-latest` to any Discussion where the user should retry th
...
+## Automated check for Issues with missing labels
+
+We have a GitHub Action (`find-issues-with-missing-labels.yml`) to find issues on our repository that are missing labels.
+Any Issues with missing labels will be put in a list in a new "error" Issue.
+
+The Action runs each week.
+
+### Apply the correct labels manually
+
+The Action will not fix any badly labeled issues.
```
Comment (HonkingGoose) [inline docs/development/issue-labeling.md:234]:
```suggestion
The Action will _not_ fix any badly labeled issues.
```

Let's emphasize the key point here.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
Some small text changes.

File: .github/workflows/find-issues-with-missing-labels.yml [hunk 11]
```diff
@@ -0,0 +1,22 @@
+name: 'Find issues with missing labels'
+
+on:
+  schedule:
+    # Run every Sunday at midnight
+    - cron: '0 0 * * 0'
+
+permissions:
+  issues: write
```
Comment (secustor) [inline .github/workflows/find-issues-with-missing-labels.yml:9]: Do not use top level permissions these are reducing the OSSF score

**Review by secustor — COMMENTED** [review]
Else LGTM, but would prefer to keep all tools in the JS/TS ecosystem for maintenance

File: .github/workflows/find-issues-with-missing-labels.yml (hunk 11 above)
Comment (ph1l1pF) [reply .github/workflows/find-issues-with-missing-labels.yml]: updated to permission on job

**Review by rarkins — APPROVED** [review]

**Pull request merged**